README.md
# Aliyun MNS Queue Driver For Laravel

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads][ico-downloads]][link-downloads]

## Install

Via Composer

``` bash
$ composer require lokielse/laravel-mns
```

## Config

Add following service providers into your providers array in `config/app.php`

``` php
Lokielse\LaravelMNS\LaravelMNSServiceProvider::class
```

Edit your `config/queue.php`, add `mns` connection

```php
'mns'        => [
	'driver'       => 'mns',
	'key'          => env('QUEUE_MNS_ACCESS_KEY'),
	'secret'       => env('QUEUE_MNS_SECRET_KEY'),
	'endpoint'     => env('QUEUE_MNS_ENDPOINT'),
	'queue'        => env('QUEUE_NAME'),
	'wait_seconds' => 30,
]
```
About [wait_seconds](https://help.aliyun.com/document_detail/35136.html)

Edit your `.env` file

```bash
QUEUE_DRIVER=mns
QUEUE_NAME=foobar-local
QUEUE_MNS_ACCESS_KEY=your_acccess_key
QUEUE_MNS_SECRET_KEY=your_secret_key
QUEUE_MNS_ENDPOINT=http://12345678910.mns.cn-hangzhou.aliyuncs.com/
```
You should update `QUEUE_MNS_ENDPOINT` to `internal endpoint` in production mode

## Usage

First create a queue and get queue endpoint at [Aliyun MNS Console](https://mns.console.aliyun.com/)

Then update `MNS_ENDPOINT` in `.env`

Push a test message to queue

```php
Queue::push(function($job){
	/**
	 * Your statments go here
	 */
	$job->delete();
});
```

Create queue listener, run command in terminal

```bash
$ php artisan queue:listen
```
## Commands
Flush MNS messages on Aliyun

```bash
$ php artisan queue:mns:flush
```

## Security

Create RAM access control at [Aliyun RAM Console](https://ram.console.aliyun.com)

1. Create a custom policy such as `AliyunMNSFullAccessFoobar`

	```
	{
	  "Version": "1",
	  "Statement": [
		{
		  "Action": "mns:*",
		  "Resource": [
			"acs:mns:*:*:*/foobar-local",
			"acs:mns:*:*:*/foobar-sandbox",
			"acs:mns:*:*:*/foobar-production"
		  ],
		  "Effect": "Allow"
		}
	  ]
	}
	```

2. Create a user for you app such as `foobar`

3. Assign the policy `AliyunMNSFullAccessFoobar` to the user `foobar`

4. Create and get the `AccessKeyId` and `AccessKeySecret` for user `foorbar`

5. update `QUEUE_MNS_ACCESS_KEY` and `QUEUE_MNS_ACCESS_SECRET` in `.env`

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Credits

- [Loki Else][link-author]
- [abrahamgreyson](https://github.com/abrahamgreyson/laravel-mns)
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/lokielse/laravel-mns.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/lokielse/laravel-mns/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/lokielse/laravel-mns.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/lokielse/laravel-mns.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/lokielse/laravel-mns.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/lokielse/laravel-mns
[link-travis]: https://travis-ci.org/lokielse/laravel-mns
[link-scrutinizer]: https://scrutinizer-ci.com/g/lokielse/laravel-mns/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/lokielse/laravel-mns
[link-downloads]: https://packagist.org/packages/lokielse/laravel-mns
[link-author]: https://github.com/lokielse
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog
All Notable changes to `laravel-mns` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


src/MNSQueue.php
<?php

namespace Lokielse\LaravelMNS;

use Aliyun\MNS\Exception\MessageNotExistException;
use Aliyun\MNS\Requests\SendMessageRequest;
use Illuminate\Contracts\Queue\Queue as QueueContract;
use Illuminate\Queue\Queue;
use Lokielse\LaravelMNS\Adaptors\MNSAdapter;

class MNSQueue extends Queue implements QueueContract
{

    /**
     * @var MNSAdapter
     */
    protected $adapter;

    /**
     * The name of default queue.
     *
     * @var string
     */
    protected $default;

    /**
     * @var null
     */
    private $waitSeconds;


    public function __construct(MNSAdapter $adapter, $queue, $waitSeconds = null)
    {
        $this->adapter     = $adapter;
        $this->default     = $queue;
        $this->waitSeconds = $waitSeconds;
    }


    /**
     * Push a new job onto the queue.
     *
     * @param string $job
     * @param mixed  $data
     * @param string $queue
     *
     * @return mixed
     */
    public function push($job, $data = '', $queue = null)
    {
        $payload = $this->createPayload($job, $data);

        return $this->pushRaw($payload, $queue);
    }


    /**
     * Push a raw payload onto the queue.
     *
     * @param string $payload
     * @param string $queue
     * @param array  $options
     *
     * @return mixed
     */
    public function pushRaw($payload, $queue = null, array $options = [ ])
    {
        $message  = new SendMessageRequest($payload);
        $response = $this->adapter->useQueue($this->getQueue($queue))->sendMessage($message);

        return $response->getMessageId();
    }


    /**
     * Push a new job onto the queue after a delay.
     *
     * @param \DateTime|int $delay
     * @param string        $job
     * @param mixed         $data
     * @param string        $queue
     *
     * @return mixed
     */
    public function later($delay, $job, $data = '', $queue = null)
    {
        $seconds  = $this->getSeconds($delay);
        $payload  = $this->createPayload($job, $data);
        $message  = new SendMessageRequest($payload, $seconds);
        $response = $this->adapter->useQueue($this->getQueue($queue))->sendMessage($message);

        return $response->getMessageId();
    }


    /**
     * Pop the next job off of the queue.
     * @link https://help.aliyun.com/document_detail/35136.html
     *
     * @param string $queue
     *
     * @return \Illuminate\Queue\Jobs\Job|null
     */
    public function pop($queue = null)
    {
        $queue = $this->getDefaultIfNull($queue);

        try {
            $response = $this->adapter->useQueue($this->getQueue($queue))->receiveMessage($this->waitSeconds);
        } catch (MessageNotExistException $e) {
            $response = null;
        }

        if ($response) {
            return new Jobs\MNSJob($this->container, $this->adapter, $queue, $response);
        }

        return null;
    }


    /**
     * 获取默认队列名（如果当前队列名为 null）。
     *
     * @param string|null $wanted
     *
     * @return string
     */
    public function getDefaultIfNull($wanted)
    {
        return $wanted ? $wanted : $this->default;
    }


    /**
     * Get the queue or return the default.
     *
     * @param  string|null $queue
     *
     * @return string
     */
    public function getQueue($queue)
    {
        return $queue ?: $this->default;
    }
}

src/Jobs/MNSJob.php
<?php

namespace Lokielse\LaravelMNS\Jobs;

use Aliyun\MNS\Responses\ReceiveMessageResponse;
use Illuminate\Container\Container;
use Illuminate\Contracts\Queue\Job as JobContract;
use Illuminate\Queue\Jobs\Job;
use Lokielse\LaravelMNS\Adaptors\MNSAdapter;

class MNSJob extends Job implements JobContract
{

    /**
     * The class name of the job.
     *
     * @var string
     */
    protected $job;

    /**
     * The queue message data.
     *
     * @var string
     */
    protected $data;

    /**
     * @var MNSAdapter
     */
    private $adapter;


    /**
     * Create a new job instance.
     *
     * @param \Illuminate\Container\Container              $container
     * @param MNSAdapter                                   $mns
     * @param string                                       $queue
     * @param \Aliyun\MNS\Responses\ReceiveMessageResponse $job
     */
    public function __construct(Container $container, MNSAdapter $mns, $queue, ReceiveMessageResponse $job)
    {
        $this->container = $container;
        $this->adapter   = $mns;
        $this->queue     = $queue;
        $this->job       = $job;
    }


    /**
     * Fire the job.
     */
    public function fire()
    {
        $body = json_decode($this->getRawBody(), true);

        if ( ! is_array($body)) {
            throw new \InvalidArgumentException("Seems it's not a Laravel enqueued job. [$body]");
        }

        $this->resolveAndFire($body);
    }


    /**
     * Get the raw body string for the job.
     *
     * @return string
     */
    public function getRawBody()
    {
        return $this->job->getMessageBody();
    }


    /**
     * Delete the job from the queue.
     */
    public function delete()
    {
        parent::delete();
        $receiptHandle = $this->job->getReceiptHandle();
        $this->adapter->deleteMessage($receiptHandle);
    }


    /**
     * Release the job back into the queue.
     *
     * @param int $delay
     */
    public function release($delay = 1)
    {
        parent::release($delay);

        if ($delay < 1) {
            $delay = 1;
        }

        $this->adapter->changeMessageVisibility($this->job->getReceiptHandle(), $delay);
    }


    /**
     * Get the number of times the job has been attempted.
     *
     * @return int
     */
    public function attempts()
    {
        return (int) $this->job->getDequeueCount();
    }


    /**
     * Get the IoC container instance.
     *
     * @return \Illuminate\Container\Container
     */
    public function getContainer()
    {
        return $this->container;
    }
}

src/Adaptors/MNSAdapter.php
<?php

namespace Lokielse\LaravelMNS\Adaptors;

use Aliyun\MNS\AsyncCallback;
use Aliyun\MNS\Client;
use Aliyun\MNS\Model\QueueAttributes;
use Aliyun\MNS\Queue;
use Aliyun\MNS\Requests\BatchDeleteMessageRequest;
use Aliyun\MNS\Requests\BatchPeekMessageRequest;
use Aliyun\MNS\Requests\BatchReceiveMessageRequest;
use Aliyun\MNS\Requests\BatchSendMessageRequest;
use Aliyun\MNS\Requests\SendMessageRequest;
use Aliyun\MNS\Responses\BatchDeleteMessageResponse;
use Aliyun\MNS\Responses\BatchPeekMessageResponse;
use Aliyun\MNS\Responses\BatchReceiveMessageResponse;
use Aliyun\MNS\Responses\BatchSendMessageResponse;
use Aliyun\MNS\Responses\ChangeMessageVisibilityResponse;
use Aliyun\MNS\Responses\GetQueueAttributeResponse;
use Aliyun\MNS\Responses\MnsPromise;
use Aliyun\MNS\Responses\PeekMessageResponse;
use Aliyun\MNS\Responses\ReceiveMessageResponse;
use Aliyun\MNS\Responses\SendMessageResponse;
use Aliyun\MNS\Responses\SetQueueAttributeResponse;

/**
 * Class MNSAdapter
 *
 * @method string getUsing()
 * @method SetQueueAttributeResponse setAttribute( QueueAttributes $attributes )
 * @method MnsPromise setAttributeAsync( QueueAttributes $attributes, AsyncCallback $callback = null )
 * @method GetQueueAttributeResponse getAttribute( QueueAttributes $attributes )
 * @method MnsPromise getAttributeAsync( QueueAttributes $attributes, AsyncCallback $callback = null )
 * @method SendMessageResponse sendMessage( SendMessageRequest $request )
 * @method MnsPromise sendMessageAsync( SendMessageRequest $request, AsyncCallback $callback = null )
 * @method PeekMessageResponse peekMessage()
 * @method MnsPromise peekMessageAsync( AsyncCallback $callback = null )
 * @method ReceiveMessageResponse receiveMessage( int $waitSeconds = null )
 * @method MnsPromise receiveMessageAsync( AsyncCallback $callback = null )
 * @method ReceiveMessageResponse deleteMessage( string $receiptHandle )
 * @method MnsPromise deleteMessageAsync( string $receiptHandle, AsyncCallback $callback = null )
 * @method ChangeMessageVisibilityResponse changeMessageVisibility( string $receiptHandle, int $visibilityTimeout )
 * @method BatchSendMessageResponse batchSendMessage( BatchSendMessageRequest $request )
 * @method MnsPromise batchSendMessageAsync( BatchSendMessageRequest $request, AsyncCallback $callback = null )
 * @method BatchReceiveMessageResponse batchReceiveMessage( BatchReceiveMessageRequest $request )
 * @method MnsPromise batchReceiveMessageAsync( BatchReceiveMessageRequest $request, AsyncCallback $callback = null )
 * @method BatchPeekMessageResponse batchPeekMessage( BatchPeekMessageRequest $request )
 * @method MnsPromise batchPeekMessageAsync( BatchPeekMessageRequest $request, AsyncCallback $callback = null )
 * @method BatchDeleteMessageResponse batchDeleteMessage( BatchDeleteMessageRequest $request )
 * @method MnsPromise batchDeleteMessageAsync( BatchDeleteMessageRequest $request, AsyncCallback $callback = null )
 */
class MNSAdapter
{

    /**
     * Aliyun MNS Client
     *
     * @var Client
     */
    private $client;

    /**
     * Aliyun MNS SDK Queue.
     *
     * @var Queue
     */
    private $queue;

    private $using;


    public function __construct(Client $client)
    {
        $this->client = $client;
    }


    /**
     * @param $method
     * @param $parameters
     *
     * @return mixed
     */
    public function __call($method, $parameters)
    {
        return call_user_func_array([ $this->queue, $method ], $parameters);
    }


    /**
     * @param string $queue
     *
     * @return self
     */
    public function useQueue($queue)
    {
        if ($this->using != $queue) {
            $this->using = $queue;
            $this->queue = $this->client->getQueueRef($queue);
        }

        return $this;
    }
}

src/Connectors/MNSConnector.php
<?php

namespace Lokielse\LaravelMNS\Connectors;

use Aliyun\MNS\Client;
use Illuminate\Queue\Connectors\ConnectorInterface;
use Lokielse\LaravelMNS\Adaptors\MNSAdapter;
use Lokielse\LaravelMNS\MNSQueue;

class MNSConnector implements ConnectorInterface
{

    /**
     * Establish a queue connection.
     *
     * @param array $config
     *
     * @return \Illuminate\Contracts\Queue\Queue
     */
    public function connect(array $config)
    {
        return new MNSQueue($this->getAdaptor($config), $config['queue'], array_get($config, 'wait_seconds'));
    }


    /**
     * @param array $config
     *
     * @return mixed
     */
    protected function getEndpoint(array $config)
    {
        return str_replace('(s)', 's', $config['endpoint']);
    }


    /**
     * @param array $config
     *
     * @return Client
     */
    protected function getClient(array $config)
    {
        return new Client($this->getEndpoint($config), $config['key'], $config['secret']);
    }


    /**
     * @param array $config
     *
     * @return MNSAdapter
     */
    protected function getAdaptor(array $config)
    {
        return new MNSAdapter($this->getClient($config));
    }
}

src/Console/MNSClearCommand.php
<?php

namespace Lokielse\LaravelMNS\Console;

use Aliyun\MNS\Client;
use Aliyun\MNS\Requests\BatchReceiveMessageRequest;
use Exception;
use Illuminate\Console\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputOption;

class MNSClearCommand extends Command
{

    /**
     * The console command name.
     *
     * @var string
     */
    protected $name = 'queue:mns:clear';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Clear MNS Queue';


    public function __construct()
    {
        parent::__construct();
    }


    /**
     * Execute the console command.
     *
     * @return void
     */
    public function fire()
    {
        $queue      = $this->argument('queue');
        $connection = $this->option('connection');

        $config = config("queue.connections.{$connection}");

        if ( ! $queue) {
            $queue = $config['queue'];
        }

        $client = new Client($config['endpoint'], $config['key'], $config['secret']);
        $queue  = $client->getQueueRef($queue);

        $hasMessage = true;

        while ($hasMessage) {

            $this->info('Peeking messages (Polling...)');

            try {
                $response = $queue->batchPeekMessage(15);
                if ($response->getMessages()) {
                    $hasMessage = true;
                } else {
                    $hasMessage = false;
                }
            } catch (Exception $e) {
                $this->info('no messages');
                break;
            }

            $response = $queue->batchReceiveMessage(new BatchReceiveMessageRequest(15, 30));

            $handles = [ ];

            /**
             * @var  \Aliyun\MNS\Model\Message $message
             */
            foreach ($response->getMessages() as $message) {
                $handles[] = $message->getReceiptHandle();
            }

            $response = $queue->batchDeleteMessage($handles);

            if ($response->isSucceed()) {
                foreach ($handles as $handle) {
                    $this->info(sprintf("The message: %s deleted success", $handle));
                }
            }
        }
    }


    /**
     * Get the console command arguments.
     *
     * @return array
     */
    protected function getArguments()
    {
        return [
            [ 'queue', InputArgument::OPTIONAL, 'The queue name' ],
        ];
    }


    /**
     * Get the console command arguments.
     *
     * @return array
     */
    protected function getOptions()
    {
        return [
            [ 'connection', 'c', InputOption::VALUE_OPTIONAL, 'The Queue connection name', 'mns' ]
        ];
    }
}


src/LaravelMNSServiceProvider.php
<?php

namespace Lokielse\LaravelMNS;

use Illuminate\Support\ServiceProvider;
use Lokielse\LaravelMNS\Connectors\MNSConnector;
use Lokielse\LaravelMNS\Console\MNSFlushCommand;

class LaravelMNSServiceProvider extends ServiceProvider
{

    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;


    public function boot()
    {
        $this->registerConnector($this->app['queue']);

        $this->commands('command.queue.mns.flush');
    }


    /**
     * Add the connector to the queue drivers.
     *
     * @return void
     */
    public function register()
    {
        $this->registerCommand();
    }


    /**
     * Register the MNS queue connector.
     *
     * @param \Illuminate\Queue\QueueManager $manager
     *
     * @return void
     */
    protected function registerConnector($manager)
    {
        $manager->addConnector('mns', function () {
            return new MNSConnector();
        });
    }


    private function registerCommand()
    {
        $this->app->singleton('command.queue.mns.flush', function () {
            return new MNSFlushCommand();
        });
    }
}


src/Console/MNSFlushCommand.php
<?php

namespace Lokielse\LaravelMNS\Console;

use Aliyun\MNS\Client;
use Aliyun\MNS\Requests\BatchReceiveMessageRequest;
use Exception;
use Illuminate\Console\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputOption;

class MNSFlushCommand extends Command
{

    /**
     * The console command name.
     *
     * @var string
     */
    protected $name = 'queue:mns:flush';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Flush MNS Queue';


    public function __construct()
    {
        parent::__construct();
    }


    /**
     * Execute the console command.
     *
     * @return void
     */
    public function fire()
    {
        $queue      = $this->argument('queue');
        $connection = $this->option('connection');

        $config = config("queue.connections.{$connection}");

        if ( ! $queue) {
            $queue = $config['queue'];
        }

        $client = new Client($config['endpoint'], $config['key'], $config['secret']);
        $queue  = $client->getQueueRef($queue);

        $hasMessage = true;

        while ($hasMessage) {

            $this->info('Peeking messages (Polling...)');

            try {
                $response = $queue->batchPeekMessage(15);
                if ($response->getMessages()) {
                    $hasMessage = true;
                } else {
                    $hasMessage = false;
                }
            } catch (Exception $e) {
                $this->info('no messages');
                break;
            }

            $response = $queue->batchReceiveMessage(new BatchReceiveMessageRequest(15, 30));

            $handles = [ ];

            /**
             * @var  \Aliyun\MNS\Model\Message $message
             */
            foreach ($response->getMessages() as $message) {
                $handles[] = $message->getReceiptHandle();
            }

            $response = $queue->batchDeleteMessage($handles);

            if ($response->isSucceed()) {
                foreach ($handles as $handle) {
                    $this->info(sprintf("The message: %s deleted success", $handle));
                }
            }
        }
    }


    /**
     * Get the console command arguments.
     *
     * @return array
     */
    protected function getArguments()
    {
        return [
            [ 'queue', InputArgument::OPTIONAL, 'The queue name' ],
        ];
    }


    /**
     * Get the console command arguments.
     *
     * @return array
     */
    protected function getOptions()
    {
        return [
            [ 'connection', 'c', InputOption::VALUE_OPTIONAL, 'The Queue connection name', 'mns' ]
        ];
    }
}
